plutus-core/plutus-core/src/PlutusCore/DeBruijn.hs

```diff
@@ -112,72 +119,125 @@ deBruijnTermM = \case
+-- A variant that does not throw an error on free variables. See 'freeIndexGrace'.
+unDeBruijnTyGrace
+    :: MonadQuote m
+    => Type NamedTyDeBruijn uni ann -> m (Type TyName uni ann)
+unDeBruijnTyGrace = flip evalStateT BM.empty . flip runReaderT (Levels 0 BM.empty) . unDeBruijnTyGraceM
+
+-- | Convert a 'Term' with 'NamedTyDeBruijn's and 'NamedDeBruijn's into a 'Term' with 'TyName's and 'Name's.
+-- A variant that does not throw an error on free variables. See 'freeIndexGrace'.
+unDeBruijnTermGrace
+    :: MonadQuote m
+    => Term NamedTyDeBruijn NamedDeBruijn uni fun ann -> m (Term TyName Name uni fun ann)
+unDeBruijnTermGrace = flip evalStateT BM.empty . flip runReaderT (Levels 0 BM.empty) . unDeBruijnTermGraceM
+
+-- | Convert a 'Program' with 'NamedTyDeBruijn's and 'NamedDeBruijn's into a 'Program' with 'TyName's and 'Name's.
+-- A variant that does not throw an error on free variables. See 'freeIndexGrace'.
+unDeBruijnProgramGrace
+    :: MonadQuote m
+    => Program NamedTyDeBruijn NamedDeBruijn uni fun ann -> m (Program TyName Name uni fun ann)
+unDeBruijnProgramGrace (Program ann ver term) = Program ann ver <$> unDeBruijnTermGrace term
+
 unDeBruijnTyM
     :: (MonadReader Levels m, MonadQuote m, AsFreeVariableError e, MonadError e m)
     => Type NamedTyDeBruijn uni ann
     -> m (Type TyName uni ann)
-unDeBruijnTyM = \case
+unDeBruijnTyM = unDeBruijnTyMWith freeIndexThrow
+
+unDeBruijnTermM
+    :: (MonadReader Levels m, MonadQuote m, AsFreeVariableError e, MonadError e m)
+    => Term NamedTyDeBruijn NamedDeBruijn uni fun ann
+    -> m (Term TyName Name uni fun ann)
+unDeBruijnTermM = unDeBruijnTermMWith freeIndexThrow
+
+unDeBruijnTyGraceM
+    :: (MonadReader Levels m, MonadQuote m, MonadState (BM.Bimap Unique Level) m)
+    => Type NamedTyDeBruijn uni ann
+    -> m (Type TyName uni ann)
+unDeBruijnTyGraceM = unDeBruijnTyMWith freeIndexGrace
+
+unDeBruijnTermGraceM
+    :: (MonadReader Levels m, MonadQuote m, MonadState (BM.Bimap Unique Level) m)
+    => Term NamedTyDeBruijn NamedDeBruijn uni fun ann
+    -> m (Term TyName Name uni fun ann)
+unDeBruijnTermGraceM = unDeBruijnTermMWith freeIndexGrace
+
+-- | Takes a "handler" function to execute when encountering free variables.
+unDeBruijnTyMWith
+    :: (MonadReader Levels m, MonadQuote m)
+    => (Index -> m Unique)
+    -> Type NamedTyDeBruijn uni ann
+    -> m (Type TyName uni ann)
+unDeBruijnTyMWith h = \case
     -- variable case
-    TyVar ann n -> TyVar ann <$> deBruijnToTyName n
+    TyVar ann n -> TyVar ann <$> deBruijnToTyName h n
     -- binder cases
     TyForall ann tn k ty ->
         -- See NOTE: [DeBruijn indices of Binders]
         declareBinder $ do
-            tn' <- deBruijnToTyName $ set index 0 tn
-            withScope $ TyForall ann tn' k <$> unDeBruijnTyM ty
+            tn' <- deBruijnToTyName h $ set index 0 tn
+            withScope $ TyForall ann tn' k <$> unDeBruijnTyMWith h ty
```

Comment:
optional: have a local `go` function that has `h` in scope, then you wouldn't need to do quite so much passing things around

Reply:
applies throughout